src/multibody/model.hxx

```diff
@@ -80,27 +80,48 @@ namespace se3
     neutralConfiguration.conservativeResize(nq);
     neutralConfiguration.tail(joint_model.nq()) = joint_model.neutralConfiguration();
     
-    // Add a the joint frame attached to itself to the frame vector - redundant information but useful.
-    addFrame(names[idx],idx,SE3::Identity(),JOINT);
-    
     // Init and add joint index to its parent subtrees.
     subtrees.push_back(IndexVector(1));
     subtrees[idx][0] = idx;
     addJointIndexToParentSubtrees(idx);
     return idx;
   }
 
+  inline FrameIndex Model::addJointFrame (const JointIndex& jidx,
+                                                int         fidx)
+  {
+    if (fidx < 0) {
+      fidx = getFrameId(names[parents[jidx]]);
+    }
+    if (fidx >= frames.size())
+      throw std::invalid_argument ("Frame not found");
+    // Add a the joint frame attached to itself to the frame vector - redundant information but useful.
+    addFrame(Frame(names[jidx],jidx,fidx,SE3::Identity(),JOINT));
+    return frames.size() - 1;
+  }
+
   inline void Model::appendBodyToJoint(const Model::JointIndex joint_index,
                                        const Inertia & Y,
-                                       const SE3 & body_placement,
-                                       const std::string & body_name)
+                                       const SE3 & body_placement)
   {
     const Inertia & iYf = Y.se3Action(body_placement);
     inertias[joint_index] += iYf;
-
-    addFrame((body_name!="")?body_name:randomStringGenerator(8), joint_index, body_placement, BODY);
     nbody++;
   }
+
+  inline FrameIndex Model::addBodyFrame (const std::string& body_name,
+                                         const JointIndex& parentJoint,
+                                         const SE3 & body_placement,
+                                               int         previousFrame)
+  {
+    if (previousFrame < 0) {
+      previousFrame = getFrameId(names[parentJoint]);
+    }
+    if (previousFrame >= frames.size())
+      throw std::invalid_argument ("Frame not found");
```

Comment:
Pinocchio design is to rather assert invalid argument than throwing errors. We might discuss this design at large, but for now, could you stick to it?
>>>>>>> assert(previousFrame >= frames.size());